### ISSUE TITLE
[BOLT]llvm-bolt crash on aarch64

### ISSUE THREAD

**Issue by HShan886** [llvm/llvm-project#62714]
hello,
when I use latest BOLT tools to optimize my binary on AArch64 platform, then get a crash stack.
```console
Either SourceMgr should be available
UNREACHABLE executed at /opt/llvm-bolt/llvm-project/llvm/lib/MC/MCContext.cpp:1044!
 #8 0x000000000090a874 llvm::MCContext::reportCommon(llvm::SMLoc, std::function<void (llvm::SMDiagnostic&, llvm::SourceMgr const*)>) /opt/llvm-bolt/llvm-project/llvm/lib/MC/MCContext.cpp:1047:16
 #9 0x000000000090a9bc llvm::MCContext::reportError(llvm::SMLoc, llvm::Twine const&) /opt/llvm-bolt/llvm-project/llvm/lib/MC/MCContext.cpp:1054:15
#10 0x0000000000859408 adjustFixupValue(llvm::MCFixup const&, llvm::MCValue const&, unsigned long, llvm::MCContext&, llvm::Triple const&, bool) /opt/llvm-bolt/llvm-project/llvm/lib/Target/AArch64/MCTargetDesc/AArch64AsmBackend.cpp:177:19
#11 0x000000000085c3e8 (anonymous namespace)::AArch64AsmBackend::applyFixup(llvm::MCAssembler const&, llvm::MCFixup const&, llvm::MCValue const&, llvm::MutableArrayRef<char>, unsigned long, bool, llvm::MCSubtargetInfo const*) const /opt/llvm-bolt/llvm-project/llvm/lib/Target/AArch64/MCTargetDesc/AArch64AsmBackend.cpp:405:27
#12 0x00000000008efa10 llvm::MCAssembler::layout(llvm::MCAsmLayout&) /opt/llvm-bolt/llvm-project/llvm/lib/MC/MCAssembler.cpp:921:7
#13 0x00000000008efa7c llvm::MCAssembler::Finish() /opt/llvm-bolt/llvm-project/llvm/lib/MC/MCAssembler.cpp:940:35
#14 0x00000000009698d8 llvm::MCObjectStreamer::finishImpl() /opt/llvm-bolt/llvm-project/llvm/lib/MC/MCObjectStreamer.cpp:937:1
#15 0x000000000094128c llvm::MCELFStreamer::finishImpl() /opt/llvm-bolt/llvm-project/llvm/lib/MC/MCELFStreamer.cpp:718:1
#16 0x0000000000994e88 llvm::MCStreamer::finish(llvm::SMLoc) /opt/llvm-bolt/llvm-project/llvm/lib/MC/MCStreamer.cpp:1018:1
#17 0x0000000000d70f8c llvm::bolt::RewriteInstance::emitAndLink() /opt/llvm-bolt/llvm-project/bolt/lib/Rewrite/RewriteInstance.cpp:3419:28
#18 0x0000000000d633c8 llvm::bolt::RewriteInstance::run() /opt/llvm-bolt/llvm-project/bolt/lib/Rewrite/RewriteInstance.cpp:775:17
#19 0x0000000000407758 main /opt/llvm-bolt/llvm-project/bolt/tools/driver/llvm-bolt.cpp:243:29
#20 0x0000ffffa81aa32c __libc_start_main (/lib64/libc.so.6+0x2432c)
#21 0x00000000004067c0 _start (/opt/llvm-bolt/llvm-project/build/bin/llvm-bolt+0x4067c0)
```
My command:
```shell
llvm-bolt exelf -o exelf.bolt -data=output.fdata -reorder-blocks=ext-tsp -reorder-functions=hfsort -split-functions -split-all-cold -split-eh -dyno-stats -update-debug-sections -v=2
```
If delete `-update-debug-sections` option from my command list, then get an assert failure message.
```console
<unknown>:0: error: fixup not sufficiently aligned
<unknown>:0: error: fixup not sufficiently aligned
```

Any suggestion will be pleasure.

**Comment by aaupov** [comment]
@yota9 – looks like a bug in AArch64 code layout, possibly relocation-related.

**Comment by Kepontry** [comment]
@HShan886 
I also encountered this issue when using options `-split-functions` and `-split-eh` together on AArch64. It is caused by the unaligned TrapFill size 1 on AArch64. When the TrapFill is applied, the address of the symbol behind it in the MCFragment is not aligned to 4 bytes. As a result, the target address of some branch instructions are not aligned and cause the assertion during emitting.

By the way, when I was preparing my patches, I noticed that treapster had fixed this problem very effectively just a week ago.

https://reviews.llvm.org/D158191

I believe this patch can solve the problem.
